### ISSUE TITLE
3.3.8 latency issues

### ISSUE THREAD

**Issue by neo22s** [koseven/koseven#275]
Hello,

**This issue is under development and requires more in deep research from myself.**

I have upgraded to 3.3.8 and immediately the app latency spiked like crazy. Caused by DB access.  Probably due to failure on cache system.

![image](https://user-images.githubusercontent.com/241725/48083045-75d58f00-e1f4-11e8-8f3e-66c5f1f3a713.png)

It's been super difficult to test since in local on pre production we are not facing this issues. Same exact code, same software version but hardware and structure it's really different.

I have used meld, to compare 3.3.7 with 3.3.8 and nothing stood out, went file by file...

Has anyone updated a PRO environment to 3.3.8?

**Comment by errotan** [comment]
I'm using the devel branch in many places without issues. What modules are enabled at your app? Do you use vendor libs with composer? Do you have profiling, caching enabled in bootstrap?

**Comment by errotan** [comment]
Are you sure that your production version is 3.3.7 and not modified? Are you using encryption functions and running in a virtual machine? If so, do you have enough entropy? (haveged daemon may help you) If you can't reproduce this in your test environment, than your only chance is to find your problem in production. There are aprox 190 commits between 3.3.7 and 3.3.8. If I where You I would try narrowing down the commit that causes this using "halving". First you apply the first 95 commit. If the problem exists, revert and try the first 48 commit. If the problem does not exists, try the first 143 commit, etc. You can narrow down to 24 commits in a few minutes which shoud be enough. Than you probably gonna find the culprit.

**Comment by Luoti** [comment]
You probably already know this, but git supports this right at hand https://git-scm.com/docs/git-bisect

**Comment by neo22s** [comment]
Yes I know I can go certain commit but we use a deploy system and the
deployment it's a pain in the ass... :( I will see what I do. Thanks all

**Comment by piotrbaczek** [comment]
Clues:
https://github.com/koseven/koseven/commit/a3096c19d9c5bc7ec0ce261ff2f9bdf80315ae94
https://github.com/koseven/koseven/commit/72af61f9618559bbbc4d7f2e81e1d67097d5ecd4

**Comment by neo22s** [comment]
Hello,

Today finally got some time for this.

@piotrbaczek I think that is the issue... since our filesystem is super slow.

This commit is wrong, since forces the usage of file_cache even if we have set another cache type.

Tomorrow I have my sys admin with my so I will be able to test this in production.

Anyway I think we should revert from

Kohana::$_files = Kohana::file_cache('Kohana::find_file()');

Back to:

Kohana::$_files = Kohana::cache('Kohana::find_file()');

**Comment by piotrbaczek** [comment]
Let's not introduce any changes to the repo without proper Unittests next time, please ;)

**Comment by piotrbaczek** [comment]
Also https://github.com/koseven/koseven/issues/24

**Comment by errotan** [comment]
Changing back to Kohana::cache prevents caching of file paths. Take a look at the flow:

1. In bootstrap.php Kohana::init() is called
2. Inside Kohana::init() >> Kohana::cache() is called to restore previously found paths into Kohana::$_files
3. Inside Kohana::cache() >> if (class_exists('Kohana_Cache')) returns false because the Kohana::modules() call is after Kohana::init() so the Cache module path is not added.
4. Kohana::cache() falls back to Kohana::file_cache() which returns an empty array, Kohana::$_files does not get populated
5. Program continues and Kohana::find_file() will be called many times which populates Kohana::$_files
6. At the end Kohana::shutdown_handler() gets called, which tries to save Kohana::$_files using Kohana::cache()
7. Because the Kohana_Cache is loaded at this point, the file paths are gonna gets saved with a cache driver (memcache, sqlite, etc.)

So basicly you won't get back what you save at the end. I'm thinking about a solution, will make PR when done.

**Comment by neo22s** [comment]
@errotan aha I understand. Tomorrow I will do a real life testing. Maybe this is not even the issue we are facing.

thanks all for the job.

**Comment by Sebjugate** [comment]
@errotan Can we use init.php of the Cache module load the files cache and trigger a repopulate of Kohana::$_files? That way, we're only working without a Cache until the Cache module is loaded (can be loaded first).

**Comment by errotan** [comment]
Actually the the current cache() method behaives badly if a module is loaded before the cache module and uses it. Data which it saves may get lost after cache module load or what it reads maybe not be correct. Originaly in kohana file_cache() was cache() this must be reverted the way it was before!

**Comment by errotan** [comment]
I mean delete the cache method and rename file_cache back to cache.

**Comment by neo22s** [comment]
Ok, today testing in Production

Immediately after changing to 3.3.8


![image](https://user-images.githubusercontent.com/241725/48542271-0845f480-e8bf-11e8-85e3-2a30f523f224.png)

Then I made a hotfix and changed file_cache() for cache() waited 5 minutes:


![image](https://user-images.githubusercontent.com/241725/48542602-e1d48900-e8bf-11e8-8c6b-8aa7d76eab7f.png)

Issue found on 3.3.8. we need to use cache() not file_cache on the core.php.

Solutions? For me to just revert the change for 3.3.9 and publish 3.3.9 also with the validation email function fixed.

**Comment by uBizzy** [comment]
For version 3.3.9 you may include also the pull request #264 that will just add JSON filed type to mysqli database module.

**Comment by neo22s** [comment]
@uBizzy we will do a fast check on PRs left to see if they can be included, but I do not want to to risk anything, 3.3.8 was a fuck up. And I am guilty. and not only fucks up others people projects it also destroys my business, we have a terrible time down which can not happen ever again.

**Comment by uBizzy** [comment]
Fair enough!

**Comment by errotan** [comment]
@neo22s if you make a patch and revert back to Kohana::cache() you essentialy break the caching functionality! This step does not make sense. An alternatice option is to introduce an another variable like $_file_cache with default to FALSE which essentialy gives you the same result like it was in 3.3.7. This can be documented and everybody whould be happy.

P.S.: talking like "3.3.8 was a fuck up" is not nice. The code works like it supposed to. Not all people have extra slow storage access like you do.

**Comment by neo22s** [comment]
@errotan  the code does not work like it's supposed to:

1 - We have a function with a breaking change that immediately affected the entire system. see https://github.com/koseven/koseven/issues/273 

2- Kohana::cache before did not store using cache from the module cache, only from file_cache, so no it was not doing what it's supposed to be doing. When I use a static like ::cache I expect to be stored in the cache system selected not in any other one, and less in a file system, none of you have a distributed service? We do and it's critical to serve our thousands of customers.

I like to be honest and it is my fault, when I say it's fucked up, it's because it is true, I do not mind accepting it. We all make mistakes, I make many, accepting them and fixing them it's what I want to do. 3.3.8 it's wrong and it's my fault since I was the one doing the release.

I should have reviewed better.

Any fix ideas are welcome but currently it's working bad and I think it's really urgent to fix it and release 3.3.9 before anyone else gets hurt.

**Comment by errotan** [comment]
Then we should vote:

1. revert file_cache() back to cache()
2. delete cache() and rename file_cache() back to cache()
3. introduce $_file_cache variable for controlling find_file() caching behaviour

I vote for 2.

**Comment by errotan** [comment]
This current behaviour of the cache() method is inconsistent during the lifetime of the application, which very likely introduce bugs. The database module only needs a small, quick patch to be configurable to use the Cache module instead of Kohana::cache() which I'm willing to make if "2." gets voted.

**Comment by neo22s** [comment]
@errotan does not matter if 2 gets voted or not, things need to be working and currently are not.

You are taking this like some short of competition.

Currently 3.3.8 it is not working and we are wasting time having it online. Not sure the aversion towards accepting this issue. There's a problem that before was not. If we revert works. 

After having a stable version as before, we can plan to change all the modules and anything related that needs to be changed so the cache system works properly in any scenario. But 3.3.8 can not be published.

Let's be pragmatic please.

**Comment by jstrobel** [comment]
I think the solution from @neo22s would be good to have a stable version for all online and think about the next release. 
Learning from that: Eventually we need more than one person (this is nothing personal @neo22s you know *g*) to make the review for a release. Any more ideas to prevent such thing in the feature?

@errotan And yes i don't think too, that the release wasn't "a fuck up" but had a major issue on @neo22s system which should 100% prevented. Why i the file system slow? Can this reproduced on a another system?
---
> Currently 3.3.8 it is not working and we are wasting time having it online. Not sure the aversion
> towards accepting this issue. There's a problem that before was not. If we revert works.
> After having a stable version as before, we can plan to change all the modules and anything
> related that needs to be changed so the cache system works properly in any scenario. But 3.3.8 
> can not be published.
---

**Comment by neo22s** [comment]
I do not know the technical part since I am far from a system
administrator. But basically it's a mounting point where files are more or
less hosted. It's a quite complex (for me) distributed and autoescalable
system based on load balancers, memcache auroradb, etc.

**Comment by Sebjugate** [comment]
@neo22s, if you have been running 33498d1b92d8f623c309426065a0e27259564ce4 in production without a3096c19d9c5bc7ec0ce261ff2f9bdf80315ae94, how are you avoiding the problem where the file cache is not correctly saved and loaded? (::shutdown_handler() saves via Cache module because it is loaded, whereas ::init() loads from file system because Cache module is not loaded)

Are you proposing to revert both these commits?

**Comment by neo22s** [comment]
@Sebjugate I have not look the code so deeply as @errotan has done and he knows better. 

What I can say as facts is that 3.3.7 has been in production for over a year working fine. Then I put 3.3.8 and problems start please see graphics. Then I change file_cache for cache and all working smooth.

So my proposal is:

1- revert file_cache loading in core.php for cache as in 3.3.7
2- release 3.3.9 which also have another bug fix for a break change on the API
3- after that when all is stable create a new issue to review in deep the cache system and make the fixes needed.

If we do not agree there's no issue, I just want to be pragmatic here and have an stable and working version live.

In my project I can extend the class and overwrite the method so does what should be doing: Storing the cache in the selected cache.

**Comment by jstrobel** [comment]
For me, the propasal sounds
good and then we have a version online which works on all plattforms (so far we know).

**Comment by Sebjugate** [comment]
@neo22s I am not disputing there is an issue, I was just curious how it wasn't still an issue with 3.3.7. I agree that releasing 3.3.9 ASAP with minimal changes to get stable behavior comparable with 3.3.7 is most important. Let's do that first.

Once that is done, we should tackle fixing this problem as I agree with @errotan that the inconsistent behavior and guaranteed cache miss on ::init() (at least as it appears to me) needs to be addressed.

**Comment by neo22s** [comment]
@Sebjugate yeah I do not take it as anything bad really, just that here looks more aggressive everything. ;)

@errotan is right if its not working good we need to fix it, but with a plan we do all together.

**Comment by Sebjugate** [comment]
@neo22s I tried this fix on my dev 3.3.8 and it increases cache misses in find_file due to the problem @errotan pointed out. I checked out 3.3.7 and I am having high cache misses as well. For me, 3.3.8 is an IMPROVEMENT in cache hits over 3.3.7. It seems to me that this caching problem has been a problem since 3.3.7 out of the box. I am not sure how this alone is fixing your production environment, but is it possible there is something else going on? Is your production environment modified in some other way (i.e. Caching module pre-loaded) that ::init() is able to pull the files cache from somewhere other than the file system?

Sorry, I'm really not trying to be controversial! Just not confident this won't cause problems for others.

**Comment by neo22s** [comment]
I have shared the latency logs.

I have not modified the init() at core.php uses same flow as described by @errotan I verified that just in case. We load modules after.

I am sure it's wrong what is in 3.3.7 for what you have described, but also at 3.3.8 it's wrong... :S

We need to redo/review the cache system.

It's weird at 3.3.7 no one in more than a year had bigger issues?

**Comment by neo22s** [comment]
Not to brag but this is the total requests last month 4,847,355 for our app. That's why it's so critical the cache system works as charm.

**Comment by piotrbaczek** [comment]
@neo22s devel branch is failing due to this revert. Can you comment or delete the unittest that makes it fail? Thank you

**Comment by neo22s** [comment]
@piotrbaczek done.

**Comment by toitzi** [comment]
@neo22s @piotrbaczek 

I think this issue is related to the "issue" i just got:

I am currently writing unittests for my Redis Cache.
If i unittest then all routes (from test `Kohana_RouteTest`) get of course stored in
my redis cache - not in file cache (cause unittesting cache methods is run before route tests and therefore as @errotan already mentioned: Inside `Kohana::cache()` class  `Kohana_Cache` is now availiable and routes are stored as they supposed to be: in my configured cache)

I discovered this because if i test my redis cache the test `test_cache_returns_false_if_cache_dnx` fails because Unittest for `Kohana_RouteTest` also only cleans the cache directory and doesn't care if another cache is configured

**Comment by neo22s** [comment]
@toitzi totally because of this. thanks ;)
